### PR TITLE
[Merged by Bors] - feat: Lemmas about the iterated shadow

### PR DESCRIPTION
**Pull request by YaelDillies**
Move the lemmas characterising `Covby` on `Finset α` from `Data.Finset.Interval` to `Data.Finset.Basic`. Golf the other proofs in `Data.Finset.Interval` and make sure the lemma names reflect whether each lemma is about `insert` or `cons`.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Combinatorics/SetFamily/Shadow.lean [hunk 1]
```diff
@@ -104,13 +107,52 @@ theorem mem_shadow_iff_insert_mem : s ∈ ∂ 𝒜 ↔ ∃ (a : _) (_ : a ∉ s)
...
 
+/-- `t ∈ ∂𝒜` iff `t` is exactly one element less than something from `𝒜` -/
+lemma mem_shadow_iff_exists_sdiff : t ∈ ∂ 𝒜 ↔ ∃ s ∈ 𝒜, t ⊆ s ∧ (s \ t).card = 1 := by
+  simp_rw [mem_shadow_iff_insert_mem, card_eq_one]
+  constructor
+  · rintro ⟨i, hi, ht⟩
+    exact ⟨insert i t, ht, subset_insert _ _, i, insert_sdiff_cancel hi⟩
+  · rintro ⟨s, hs, hts, a, ha⟩
+    refine' ⟨a, (mem_sdiff.1 $ (ext_iff.1 ha _).2 $ mem_singleton_self _).2, _⟩
+    rwa [insert_eq, ←ha, sdiff_union_of_subset hts]
```
Comment (eric-wieser) [inline Mathlib/Combinatorics/SetFamily/Shadow.lean:118]: Is there an `sdiff_card_eq_one` lemma hiding here, with conclusion ∃ (a : _) (_ : a ∉ t), insert a t = s`?

Comment (YaelDillies) [reply Mathlib/Combinatorics/SetFamily/Shadow.lean:118]: There's a bunch of equivalences here, eg `t = s.erase a`, `s` covers `t`... Do I need a TFAE statement?

Comment (eric-wieser) [reply Mathlib/Combinatorics/SetFamily/Shadow.lean:118]: I think the covby statement is the right one for. `card_sdiff_eq_one`, good point.

File: Mathlib/Combinatorics/SetFamily/Shadow.lean [hunk 2]
```diff
@@ -104,13 +107,52 @@ theorem mem_shadow_iff_insert_mem : s ∈ ∂ 𝒜 ↔ ∃ (a : _) (_ : a ∉ s)
     exact ⟨insert a s, hs, a, mem_insert_self _ _, erase_insert ha⟩
 #align finset.mem_shadow_iff_insert_mem Finset.mem_shadow_iff_insert_mem
 
+/-- `t ∈ ∂𝒜` iff `t` is exactly one element less than something from `𝒜` -/
```
Comment (b-mehta) [inline Mathlib/Combinatorics/SetFamily/Shadow.lean:110]: Perhaps a docstring referencing the very similar lemma `mem_shadow_iff_exists_mem_card_add_one` below would be nice (and similar for iterated). I expect one can be used to golf the other?

File: Mathlib/Combinatorics/SetFamily/Shadow.lean (hunk 1 above)
Comment (YaelDillies) [reply Mathlib/Combinatorics/SetFamily/Shadow.lean]: This worked out nicely.

File: Mathlib/Combinatorics/SetFamily/Shadow.lean [hunk 3]
```diff
@@ -93,24 +96,79 @@ theorem erase_mem_shadow (hs : s ∈ 𝒜) (ha : a ∈ s) : erase s a ∈ ∂ 
...
 /-- `t` is in the shadow of `𝒜` iff we can add an element to it so that the resulting finset is in
 `𝒜`. -/
-theorem mem_shadow_iff_insert_mem : s ∈ ∂ 𝒜 ↔ ∃ (a : _) (_ : a ∉ s), insert a s ∈ 𝒜 := by
-  refine' mem_shadow_iff.trans ⟨_, _⟩
-  · rintro ⟨s, hs, a, ha, rfl⟩
-    refine' ⟨a, not_mem_erase a s, _⟩
-    rwa [insert_erase ha]
-  · rintro ⟨a, ha, hs⟩
-    exact ⟨insert a s, hs, a, mem_insert_self _ _, erase_insert ha⟩
+theorem mem_shadow_iff_insert_mem : s ∈ ∂ 𝒜 ↔ ∃ a, a ∉ s ∧ insert a s ∈ 𝒜 := by
```
Comment (b-mehta) [inline Mathlib/Combinatorics/SetFamily/Shadow.lean:107]: Nit: I don't like that some lemmas use `s` for things in the shadow, and others use it for things in the original set

File: Mathlib/Combinatorics/SetFamily/Shadow.lean [hunk 4]
```diff
@@ -93,24 +96,79 @@ theorem erase_mem_shadow (hs : s ∈ 𝒜) (ha : a ∈ s) : erase s a ∈ ∂ 
...
+  constructor
+  · rintro ⟨t, ⟨a, ha, hs⟩, hts, rfl⟩
+    refine' ⟨_, hs, hts.trans $ subset_insert _ _, _⟩
+    rw [insert_sdiff_of_not_mem _ $ not_mem_mono hts ha,
+      card_insert_of_not_mem $ not_mem_mono (sdiff_subset _ _) ha]
+  · rintro ⟨t, hs, hts, hk⟩
+    obtain ⟨a, ha⟩ : (t \ s).Nonempty := by rw [←card_pos, hk]; exact Nat.succ_pos _
+    refine' ⟨erase t a, ⟨a, not_mem_erase _ _, _⟩, subset_erase.2 ⟨hts, (mem_sdiff.1 ha).2⟩, _⟩
+    · rwa [insert_erase (mem_sdiff.1 ha).1]
+    · rw [erase_sdiff_comm, card_erase_of_mem ha, hk, succ_sub_one]
```
Comment (b-mehta) [inline Mathlib/Combinatorics/SetFamily/Shadow.lean:145]:
```suggestion

lemma mem_shadow_iterate_iff_exists_third :
    s ∈ ∂^[k] 𝒜 ↔ ∃ u : Finset α, u.card = k ∧ Disjoint s u ∧ s ∪ u ∈ 𝒜 := by
  induction' k with k ih generalizing s
  · simp
  simp only [mem_shadow_iff_insert_mem, ih, Function.iterate_succ_apply', card_eq_succ]
  aesop

/-- `t ∈ ∂^k 𝒜` iff `t` is exactly `k` elements less than something from `𝒜`.

See also `Finset.mem_shadow_iff_exists_mem_card_add`. -/
lemma mem_shadow_iterate_iff_exists_sdiff :
    s ∈ ∂^[k] 𝒜 ↔ ∃ t ∈ 𝒜, s ⊆ t ∧ (t \ s).card = k := by
  rw [mem_shadow_iterate_iff_exists_third]
  constructor
  · rintro ⟨u, rfl, hsu, hsuA⟩
    exact ⟨_, hsuA, subset_union_left s u, by rw [union_sdiff_cancel_left hsu]⟩
  rintro ⟨t, htA, hst, rfl⟩
  refine ⟨t \ s, rfl, disjoint_sdiff, ?_⟩
  rwa [union_sdiff_self_eq_union, union_eq_right.2 hst]
```
Here's a way to refactor which makes the induction part a bit clearer, gives you another useful-looking statement generalising `mem_shadow_iff_insert_mem`, and makes the proof of `mem_shadow_iterate_iff_exists_sdiff` simpler. Feel free to change the name though

File: Mathlib/Data/Finset/Card.lean [hunk 5]
```diff
@@ -656,6 +657,27 @@ theorem card_eq_three [DecidableEq α] :
...
+    simp [*]
+  tfae_have 2 → 3
+  · simp_rw [card_eq_one]
+    rintro ⟨hts, a, ha⟩
+    refine ⟨a, (mem_sdiff.1 $ superset_of_eq ha $ mem_singleton_self _).2, ?_⟩
+    rw [insert_eq, ←ha, sdiff_union_of_subset hts]
+  tfae_finish
+
+lemma covby_iff_card_sdiff_eq_one : t ⋖ s ↔ t ⊆ s ∧ (s \ t).card = 1 := covby_tfae.out 0 1
+
```
Comment (eric-wieser) [inline Mathlib/Data/Finset/Card.lean:678]: I'm not convinced this TFAE is useful; you could achieve the same thing with something along the lines of
```suggestion
lemma covby_iff_card_sdiff_eq_one : t ⋖ s ↔ t ⊆ s ∧ (s \ t).card = 1 := by
  rw [covby_iff_exists_insert]
  constructor 
  · rintro ⟨a, ha, rfl⟩
    simp [*]
  · simp_rw [card_eq_one]
    rintro ⟨hts, a, ha⟩
    refine ⟨a, (mem_sdiff.1 $ superset_of_eq ha $ mem_singleton_self _).2, ?_⟩
    rw [insert_eq, ←ha, sdiff_union_of_subset hts]
```

Comment (YaelDillies) [reply Mathlib/Data/Finset/Card.lean:678]: But surely that TFAE is useful if only to say  "This is all the equivalences we have in the library"?

Comment (b-mehta) [reply Mathlib/Data/Finset/Card.lean]: I agree we don't have 1 or 2 here, but this TFAE could still be useful in other contexts where going between any of this four concepts is useful. Though we do seemingly have a canonical spelling for this, so perhaps just the `covby_iff_` lemmas suffice.

**Comment by b-mehta** [comment]
Looks good to me, save for the TFAE discussion, which I'm neutral on.

**Review by b-mehta — APPROVED** [review]

File: Mathlib/Data/Finset/Card.lean (hunk 5 above)
Comment (YaelDillies) [reply Mathlib/Data/Finset/Card.lean]: I've dumped the TFAE for what it's worth.

**Comment by b-mehta** [comment]
bors merge

**Comment by ocfnash** [comment]
GitHub was flaking and bors crashed so we need to try this again

bors merge

**Pull request closed without merging**